Sources/Runestone/TextView/TextInput/LayoutManager.swift

```diff
@@ -420,11 +420,17 @@ extension LayoutManager {
         let safeLocation = min(max(location, 0), stringView.string.length)
         let line = lineManager.line(containingCharacterAt: safeLocation)!
         let lineController = lineControllerStorage.getOrCreateLineController(for: line)
-        let localLocation = safeLocation - line.location
-        let localCaretRect = lineController.caretRect(atIndex: localLocation)
-        let globalYPosition = line.yPosition + localCaretRect.minY
-        let globalRect = CGRect(x: localCaretRect.minX, y: globalYPosition, width: localCaretRect.width, height: localCaretRect.height)
-        return globalRect.offsetBy(dx: leadingLineSpacing, dy: textContainerInset.top)
+        let lineLocalLocation = safeLocation - line.location
+        let lineFragmentNode = lineController.lineFragmentNode(containingCharacterAt: lineLocalLocation)
```

Comment:
I'm getting assert failures from this line:

### Case 1

In my particular use of Runestone queries `TextView.rect(of:)` a lot to anchor an autocompletion view to the text view.

My implementation ends up calling this right after loading the text view, with the `location` being the very end of the text. This line ends up calling `RedBlackTree.node(containingLocation:)` which throws this assert:

```
assert(location <= root.nodeTotalValue)
```

At that time `root.nodeTotalValue` is 0, so the assert fires. Without the assert the next line in that method will crash due to the force-unwrap.

### Case 2

The same assertion failure fires when I have a line that's split into line fragments, and I then try to insert a line break in a line fragment by pressing "Enter".

`location` is 1 and `root.nodeTotalValue` is 0.

Stack trace:

```
#0	0x000000019613132c in _swift_runtime_on_report ()
#1	0x00000001961c68d8 in _swift_stdlib_reportFatalErrorInFile ()
#2	0x0000000195de958c in _assertionFailure(_:_:file:line:flags:) ()
#3	0x000000010ded6dc0 in RedBlackTree.node(containingLocation:) at (DIR)/forks/Runestone/Sources/Runestone/RedBlackTree/RedBlackTree.swift:39
#4	0x000000010def5d48 in LineController.lineFragmentNode(containingCharacterAt:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/LineController/LineController.swift:168
#5	0x000000010df3920c in LayoutManager.caretRect(at:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/LayoutManager.swift:424
#6	0x000000010df3da50 in LayoutManager.getLineSelectionRect() at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/LayoutManager.swift:596
#7	0x000000010df3bc38 in LayoutManager.layoutLineSelection() at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/LayoutManager.swift:568
#8	0x000000010df2fd38 in LayoutManager.layoutIfNeeded() at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/LayoutManager.swift:537
#9	0x000000010df6c7dc in TextInputView.applyLineChangesToLayoutManager(_:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/TextInputView.swift:1164
#10	0x000000010df69da0 in TextInputView.replaceText(in:with:selectedRangeAfterUndo:undoActionName:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/TextInputView.swift:1141
#11	0x000000010df712e8 in TextInputView.indentController(_:shouldInsert:in:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/TextInputView.swift:1551
#12	0x000000010df71514 in protocol witness for IndentControllerDelegate.indentController(_:shouldInsert:in:) in conformance TextInputView ()
#13	0x000000010df24f64 in IndentController.insertLineBreak(in:using:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/IndentController.swift:134
#14	0x000000010df693d4 in TextInputView.insertText(_:) at (DIR)/forks/Runestone/Sources/Runestone/TextView/TextInput/TextInputView.swift:995
```

Reply:
(In case it matters: I was testing this on Mac Catalyst)

Reply:
Thanks for checking this and for your detailed report. I was unable to reproduce the issue in a Mac Catalyst app but I've attempted to fix it in https://github.com/simonbs/Runestone/pull/153/commits/07cf17703bf5615402d02f94379decdaebf9e203. Can you check if you're still seeing the issue?

Reply:
I've tested this a bunch and it worked _flawlessly_. Much appreciated!

Reply:
@nighthawk Glad to hear that! Thanks a lot for testing this and your feedback along the way 😃